File: ideaproject/javaserializer/src/com/houseelectrics/serializer/path/PathUtil.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using com.houseelectrics.serializer;

namespace com.houseelectrics.serializer.path
{
    public class InstancePathElement
    {
        public bool IsIndexed { get; set; }
    }

    public class NamePathElement : InstancePathElement
        {
        public override string ToString()
        {
            return Name;
        }
            public string Name { get; set; }
        }
    public class IndexPathElement : InstancePathElement
        {
            public int Index { get; set; }
            public override string ToString()
            {
                return "[" + Index + "]";
            }    
    }

    public class PathUtil
    {
        public List<InstancePathElement> findSingleObjectPath(object root, object object2Find)
        {
            bool found=false;
            ObjectExplorerImpl explorer = new ObjectExplorerImpl();
            Stack<InstancePathElement> currentPath = new Stack<InstancePathElement>();

            MoveAway down = delegate(Object from, string propertyName, Object to, bool isIndexed, int? index)
{
    if (found) return false;
    InstancePathElement element;
    bool parentIsIndexed = currentPath.Count > 0 && currentPath.Peek().IsIndexed;
    if (parentIsIndexed)
    {
        IndexPathElement ielement = new IndexPathElement();
        ielement.Index = (int)index;
        element = ielement;
    }
    else
    {
        NamePathElement nElement = new NamePathElement();
        nElement.Name = propertyName;
        element = nElement;
    }
    element.IsIndexed = isIndexed;
    currentPath.Push(element);
    found = object2Find == to;
    return (!found);
};
MoveBack up = (from, propertyName, to, isIndexed) =>
{
    if (!found) currentPath.Pop();
};

OnLeaf onLeaf = (from, propertyName, to, index) =>
    {
    };

    PropertyReflectionNodeExpander expander = new PropertyReflectionNodeExpander();
    expander.ExcludeReadOnlyProperties = true;
    explorer.NodeExpander = expander;

    explorer.explore(root, down, up, onLeaf);
    List<InstancePathElement> result = null; 
            if (found)
            {
                result = new List<InstancePathElement>(currentPath.ToArray().Reverse());
                // remove the first element because it is to the root
                result.RemoveAt(0);
            }
            return result;
        }

        public string ToString(List<InstancePathElement> path)
{
    StringBuilder sb = new StringBuilder();
    for (int done =0 ; done<path.Count; done++)
    {
        if (done != 0) sb.Append('.');
        sb.Append(path[done]);
    }
    return sb.ToString();
}
    
    }



}

File: ideaproject/javaserializer/src/com/houseelectrics/serializer/path/PathUtil.java
package com.houseelectrics.serializer.path;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import com.houseelectrics.serializer.ExplorationListener;
import com.houseelectrics.serializer.ObjectExplorerImpl;
import com.houseelectrics.serializer.PropertyReflectionNodeExpander;

/**
 * Created by roberttodd on 04/12/2014.
 */
public class PathUtil
{

    public static class InstancePathElement
    {
        private boolean isIndexed;
        public boolean getIsIndexed() {return isIndexed;}
        public void setIsIndexed(boolean value) {this.isIndexed = value;}
    }

    public static class NamePathElement extends InstancePathElement
    {
        public String toString()
        {
            return name;
        }
        private String name;
        public String getName() {return name;}
        public void setName(String value){this.name = value; }
    }
    public static class IndexPathElement extends InstancePathElement
    {
        private int index;
        public int getIndex() { return index; }
        public void setIndex(int value) {this.index = value;}
    public String toString()
{
    return "[" + getIndex() + "]";
}
}

    private boolean found;
    public List<InstancePathElement> findSingleObjectPath(Object root, final Object object2Find)
    {
        this.found=false;
        ObjectExplorerImpl explorer = new ObjectExplorerImpl();
        final Stack<InstancePathElement> currentPath = new Stack<InstancePathElement>();

        ExplorationListener explorationListener = new ExplorationListener()
        {
            @Override
            public boolean MoveAway(Object from, String propertyName, Object to, boolean isIndexed, Integer index)
            {
                if (found) return false;
                InstancePathElement element;
                InstancePathElement thepeek = currentPath.size() > 0 ? currentPath.peek() : null;
                boolean parentIsIndexed = currentPath.size() > 0 && currentPath.peek().getIsIndexed();
                if (parentIsIndexed)
                {
                    IndexPathElement ielement = new IndexPathElement();
                    ielement.setIndex( index.intValue());
                    element = ielement;
                }
                else
                {
                    NamePathElement nElement = new NamePathElement();
                    nElement.setName( propertyName);
                    element = nElement;
                }
                element.setIsIndexed(isIndexed);
                currentPath.push(element);
                found = object2Find == to;
                return (!found);
            }

            @Override
            public void MoveBack(Object from, String propertyName, Object to, boolean isIndexed)
            {
                if (!found) currentPath.pop();
            }

            @Override
            public void OnLeaf(Object from, String propertyName, Object to, Integer index)
            {
            }
        };


        PropertyReflectionNodeExpander expander = new PropertyReflectionNodeExpander();
        expander.setExcludeReadOnlyProperties( true );
        explorer.setNodeExpander(expander);

        explorer.explore(root, explorationListener);
        List<InstancePathElement> result = null;
        if (found)
        {
            result = new ArrayList<InstancePathElement>(currentPath.size());
            for (int index = 0; index<currentPath.size(); index++)
            {
                result.add(currentPath.get(index));
            }
            // remove the first element because it is to the root
            result.remove(0);
        }
        return result;
    }

    public String toString(List<InstancePathElement> path)
    {
        StringBuilder sb = new StringBuilder();
        for (int done =0 ; done<path.size(); done++)
        {
            if (done != 0) sb.append('.');
            sb.append(path.get(done));
        }
        return sb.toString();
    }

}
